File: ocp-python/src/ocp_agent/registry.py
"""
OCP Registry Client

Provides access to the OCP Community Registry for pre-discovered API specifications.
Enables fast API registration without OpenAPI spec parsing.
"""

import os
import requests
from typing import List, Optional, Dict, Any
from urllib.parse import urljoin

from .schema_discovery import OCPAPISpec, OCPTool
from .errors import RegistryUnavailable, APINotFound


class OCPRegistry:
    """
    Registry client for pre-discovered API specifications.
    
    Provides fast API lookup from the OCP Community Registry, which stores
    pre-parsed OpenAPI specifications as ready-to-use tool collections.
    """
    
    def __init__(self, registry_url: Optional[str] = None):
        """
        Initialize registry client.
        
        Args:
            registry_url: Registry base URL. Falls back to OCP_REGISTRY_URL env var,
                         then default public registry.
        """
        self.registry_url = (
            registry_url or 
            os.getenv('OCP_REGISTRY_URL', 'https://registry.ocp.dev')
        ).rstrip('/')
        
        # Validate URL format
        if not self.registry_url.startswith(('http://', 'https://')):
            raise ValueError(f"Invalid registry URL: {self.registry_url}")
    
    def get_api_spec(self, name: str, base_url: Optional[str] = None) -> OCPAPISpec:
        """
        Get pre-discovered API specification from registry.
        
        Args:
            name: API name in registry (e.g., 'github', 'stripe')
            base_url: Override base URL for API calls
            
        Returns:
            OCPAPISpec with pre-discovered tools
            
        Raises:
            RegistryUnavailable: Registry service unreachable
            APINotFound: API not found in registry
        """
        try:
            # Get API entry from registry
            response = requests.get(
                urljoin(self.registry_url, f'/api/v1/registry/{name}'),
                timeout=10
            )
            
            if response.status_code == 404:
                # Try to get suggestions
                suggestions = self._get_suggestions(name)
                raise APINotFound(name, suggestions)
            
            response.raise_for_status()
            api_entry = response.json()
            
            # Convert registry entry to OCPAPISpec
            return self._entry_to_spec(api_entry, base_url)
            
        except requests.exceptions.RequestException as e:
            raise RegistryUnavailable(self.registry_url, str(e))
    
    def search_apis(self, query: str) -> List[str]:
        """
        Search available APIs for suggestions.
        
        Args:
            query: Search query
            
        Returns:
            List of API names matching query
        """
        try:
            response = requests.get(
                urljoin(self.registry_url, '/api/v1/search'),
                params={'q': query, 'per_page': 10},
                timeout=5
            )
            response.raise_for_status()
            
            search_results = response.json()
            return [result['name'] for result in search_results.get('results', [])]
            
        except requests.exceptions.RequestException:
            # If search fails, return empty list (non-critical)
            return []
    
    def list_apis(self) -> List[str]:
        """
        List all available APIs in registry.
        
        Returns:
            List of API names
        """
        try:
            response = requests.get(
                urljoin(self.registry_url, '/api/v1/registry'),
                timeout=10
            )
            response.raise_for_status()
            
            apis = response.json()
            return [api['name'] for api in apis]
            
        except requests.exceptions.RequestException:
            return []
    
    def _entry_to_spec(self, api_entry: Dict[str, Any], base_url_override: Optional[str] = None) -> OCPAPISpec:
        """Convert registry API entry to OCPAPISpec."""
        
        # Use override base_url if provided, otherwise use registry's base_url
        base_url = base_url_override or api_entry.get('base_url', '')
        
        # Convert tools from registry format to OCPTool objects
        tools = []
        if api_entry.get('tools'):
            for tool_dict in api_entry['tools']:
                tool = OCPTool(
                    name=tool_dict['name'],
                    description=tool_dict['description'],
                    method=tool_dict['method'],
                    path=tool_dict['path'],
                    parameters=tool_dict.get('parameters', {}),
                    response_schema=tool_dict.get('response_schema', {}),
                    operation_id=tool_dict.get('operation_id'),
                    tags=tool_dict.get('tags', [])
                )
                tools.append(tool)
        
        # Create OCPAPISpec (matching dataclass field order)
        return OCPAPISpec(
            base_url=base_url,
            title=api_entry.get('display_name', api_entry['name']),
            version='1.0.0',  # Registry doesn't store version
            description=api_entry.get('description', ''),
            tools=tools,
            raw_spec=api_entry  # Store the original registry entry
        )
    
    def _get_suggestions(self, api_name: str) -> List[str]:
        """Get API name suggestions for error messages."""
        # Try exact search first
        suggestions = self.search_apis(api_name)
        
        # If no results, try partial matches
        if not suggestions and len(api_name) > 2:
            suggestions = self.search_apis(api_name[:3])
        
        return suggestions[:3]  # Limit to 3 suggestions

File: ocp-javascript/src/registry.ts
/**
 * OCP Registry Client
 * 
 * Client for accessing pre-discovered API specifications from the OCP registry.
 */

import { OCPAPISpec, OCPTool } from './schema_discovery.js';
import { RegistryUnavailable, APINotFound } from './errors.js';

const DEFAULT_REGISTRY_URL = 'https://registry.ocp.dev';

/**
 * Registry API entry structure
 */
interface RegistryAPIEntry {
    name: string;
    display_name?: string;
    description?: string;
    base_url?: string;
    tools?: Array<{
        name: string;
        description: string;
        method: string;
        path: string;
        parameters?: Record<string, any>;
        response_schema?: Record<string, any>;
        operation_id?: string;
        tags?: string[];
    }>;
}

/**
 * OCP Registry Client
 * 
 * Access pre-discovered API specifications from the OCP community registry.
 */
export class OCPRegistry {
    private registryUrl: string;

    /**
     * Initialize registry client.
     * 
     * @param registryUrl - Registry URL (defaults to OCP_REGISTRY_URL env var or registry.ocp.dev)
     */
    constructor(registryUrl?: string) {
        this.registryUrl = (registryUrl || process.env.OCP_REGISTRY_URL || DEFAULT_REGISTRY_URL).replace(/\/$/, '');
        
        // Validate URL format
        if (!this.registryUrl.startsWith('http://') && !this.registryUrl.startsWith('https://')) {
            throw new Error(`Invalid registry URL: ${this.registryUrl}`);
        }
    }

    /**
     * Get API specification from registry.
     * 
     * @param name - API name in registry
     * @param baseUrl - Optional override for API base URL
     * @returns API specification with tools
     * @throws RegistryUnavailable if registry is unreachable
     * @throws APINotFound if API is not found in registry
     */
    async getApiSpec(name: string, baseUrl?: string): Promise<OCPAPISpec> {
        try {
            const response = await fetch(`${this.registryUrl}/api/v1/registry/${name}`, {
                timeout: 10000
            } as any);
            
            if (response.status === 404) {
                const suggestions = await this._getSuggestions(name);
                throw new APINotFound(name, suggestions);
            }
            
            if (!response.ok) {
                throw new RegistryUnavailable(
                    this.registryUrl,
                    `${response.statusText}`
                );
            }
            
            const entry = await response.json() as RegistryAPIEntry;
            return await this._entryToSpec(entry, baseUrl);
            
        } catch (error) {
            if (error instanceof APINotFound || error instanceof RegistryUnavailable) {
                throw error;
            }
            
            if (error instanceof TypeError && error.message.includes('fetch')) {
                throw new RegistryUnavailable(
                    this.registryUrl,
                    `Could not connect to registry`
                );
            }
            
            throw new RegistryUnavailable(
                this.registryUrl,
                String(error)
            );
        }
    }

    /**
     * Search for APIs in the registry.
     * 
     * @param query - Search query string
     * @returns List of API names matching query
     */
    async searchApis(query: string): Promise<string[]> {
        try {
            const params = new URLSearchParams({ 
                q: query,
                per_page: '10'
            });
            
            const url = `${this.registryUrl}/api/v1/search?${params.toString()}`;
            const response = await fetch(url, {
                timeout: 5000
            } as any);
            
            if (!response.ok) {
                return [];
            }
            
            const searchResults = await response.json() as { results?: Array<{ name: string }> };
            return (searchResults.results || []).map(r => r.name);
            
        } catch {
            // If search fails, return empty list (non-critical)
            return [];
        }
    }

    /**
     * List all APIs in the registry.
     * 
     * @returns List of API names
     */
    async listApis(): Promise<string[]> {
        try {
            const response = await fetch(`${this.registryUrl}/api/v1/registry`, {
                timeout: 10000
            } as any);
            
            if (!response.ok) {
                return [];
            }
            
            const apis = await response.json() as Array<{ name: string }>;
            return apis.map(api => api.name);
            
        } catch {
            // If list fails, return empty list (non-critical)
            return [];
        }
    }

    /**
     * Convert registry entry to API spec.
     */
    private async _entryToSpec(entry: RegistryAPIEntry, baseUrlOverride?: string): Promise<OCPAPISpec> {
        // Use override base_url if provided, otherwise use registry's base_url
        const baseUrl = baseUrlOverride || entry.base_url || '';
        
        // Convert tools from registry format to OCPTool objects
        const tools: OCPTool[] = [];
        if (entry.tools) {
            for (const toolDict of entry.tools) {
                const tool: OCPTool = {
                    name: toolDict.name,
                    description: toolDict.description,
                    method: toolDict.method,
                    path: toolDict.path,
                    parameters: toolDict.parameters || {},
                    response_schema: toolDict.response_schema || {},
                    operation_id: toolDict.operation_id,
                    tags: toolDict.tags || []
                };
                tools.push(tool);
            }
        }
        
        // Create OCPAPISpec (matching dataclass field order)
        return {
            base_url: baseUrl,
            title: entry.display_name || entry.name,
            version: '1.0.0',  // Registry doesn't store version
            description: entry.description || '',
            tools,
            raw_spec: entry  // Store the original registry entry
        };
    }

    /**
     * Get API name suggestions for error messages.
     */
    private async _getSuggestions(apiName: string): Promise<string[]> {
        try {
            // Try exact search first
            let suggestions = await this.searchApis(apiName);
            
            // If no results, try partial matches
            if (suggestions.length === 0 && apiName.length > 2) {
                suggestions = await this.searchApis(apiName.slice(0, 3));
            }
            
            return suggestions.slice(0, 3);
            
        } catch {
            return [];
        }
    }
}
